### PR TITLE
Feat/accountview

### PR DESCRIPTION
**Pull request by crypblizz8**
Demo

Overview
- Implement the UI Button for `AccountButton` and `AccountView`
- AccountButton displays: `balance`, `ensAvatar` or `Zorb`, `address`
- AccountView displays: same features as above + `disconnect` + `copyClipboard` functions

Considerations
- In `getENSAvatar()` and `getBalance()`, should look to add them to AccountCtrl so it is only fetched once. However this may apply to other functions throughout the codeBase so needs discussion.
- Consider loading states / completed state for `copyClipboard` + ens/zorb Avatars

Other
- Long commit list as I had to rebase + move my other PR here.

### CODE REVIEW

File: packages/ui/src/components/w3m-modal-router/index.ts [hunk 1]
```diff
@@ -34,8 +35,8 @@ export class W3mModalRouter extends LitElement {
   }
 
   public firstUpdated() {
-    this.resizeObserver = new ResizeObserver(([conetnt]) => {
-      const newHeight = `${conetnt.borderBoxSize[0].blockSize}px`
+    this.resizeObserver = new ResizeObserver(([content]) => {
```
Comment (IljaDaderko) [inline packages/ui/src/components/w3m-modal-router/index.ts:38]: Nice catch!

File: packages/ui/src/components/w3m-zorb-ens-image/styles.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+export default css`
+  .w3m-ens-large-image {
+    overflow: hidden;
+    position: relative;
+    border-radius: 30px;
+    width: 60px;
+    height: 60px;
+    margin-bottom: 10px;
+  }
```
Comment (IljaDaderko) [inline packages/ui/src/components/w3m-zorb-ens-image/styles.ts:11]: Could you add line break between css blocks for better readability please

File: packages/ui/src/components/w3m-zorb-ens-image/index.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  // -- state & properties ------------------------------------------- //
+  @property() public address = ''
+  @property() public size = ''
+  @property() public ens = ''
+
+  // -- private ----------------------------------------------------- //
+  private async getENSAvatar() {
+    try {
+      const opts: FetchEnsAvatarOpts = {
```
Comment (IljaDaderko) [inline packages/ui/src/components/w3m-zorb-ens-image/index.ts:27]: You shouldn't really need to specify `: FetchEnsAvatarOpts` here, as we try to keep everything to typescripts auto-inheritance i.e. if it is wrong it should be highlighted on `.fetchEnsAvatar(opts)` level

File: packages/ui/src/partials/w3m-account-button/index.ts [hunk 4]
```diff
@@ -1,16 +1,100 @@
...
+
+  // -- state & properties ------------------------------------------- //
+  @state() private address = ''
+  @state() private balance = ''
+  @state() private ens = ''
+
+  // -- lifecycle ---------------------------------------------------- //
+  public constructor() {
+    super()
+    this.getAccounts()
```
Comment (IljaDaderko) [inline packages/ui/src/partials/w3m-account-button/index.ts:26]: This one is gonna be tricky, but I'll try to outline proper setps best I can:

1. Ideally we should have all of this data (acc, ens and balance) available here before the component is mounted. This can be done inside `chains/ethereum/src/utilities` `onClientConnected` function. And should also happen in `Promise.all`

2. Doing it in the above way also means that we probably need to move all `ConnectModalCtrl.closeModal` calls to `onClientConnected` as well, so that it closes after all the data is ready

3. Inside w3m-account-button component you would then need to not only access the data, but also subscribe to it (account can change if user changes it on their mobile wallet for example). You can see example of subscribing and unsubscribing to state in component like `w3m-modal-router`

File: packages/ui/src/partials/w3m-account-button/styles.ts [hunk 5]
```diff
@@ -60,12 +118,28 @@ export function dynamicStyles() {
       color: ${foreground[3]};
     }
 
-    svg path {
-      fill: ${foreground.inverse};
-    }
+    // svg path {
```
Comment (IljaDaderko) [inline packages/ui/src/partials/w3m-account-button/styles.ts:121]: Will this be un-commented latter or can it be removed? If it is used latter on, could you add TODO here please

File: packages/ui/src/utils/Svgs.ts [hunk 6]
```diff
@@ -451,3 +451,77 @@ export const DEFI_IMG = svg`
     </defs>
   </svg>
 `
+
+export const ETH_LOGO = svg`
```
Comment (IljaDaderko) [inline packages/ui/src/utils/Svgs.ts:455]: For all new svg's could you please run them through https://jakearchibald.github.io/svgomg/ and cleanup unecessary data i.e. `fill="none" xmlns="http://www.w3.org/2000/svg"` as these are inlined into final bundle.

File: packages/ui/src/utils/UiHelpers.ts [hunk 7]
```diff
@@ -63,3 +63,27 @@ export async function preloadImage(src: string) {
...
+  script.async = true
+
+  document.body.appendChild(script)
+
+  return () => {
+    document.body.removeChild(script)
+  }
+}
+
+export const ZorbPackageScript =
```
Comment (IljaDaderko) [inline packages/ui/src/utils/UiHelpers.ts:80]: Any 3rd party lib we use, should come from npm. If it is not available there, you can copy source to utils i.e `Zorb.ts` (see how this was done for Whatmesh). And then import it where it is required (see if you can strip it down as well for unused code). This means that `useScript` util will also not be required.

File: packages/ui/src/views/w3m-account-view/index.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+  // -- state & properties ------------------------------------------- //
+  @state() private address = ''
+  @state() private balance = ''
+  @state() private ens = ''
+
+  // -- lifecycle ---------------------------------------------------- //
+  public constructor() {
+    super()
+    this.getAccounts()
```
Comment (IljaDaderko) [inline packages/ui/src/views/w3m-account-view/index.ts:26]: Same as above. This data should be available beforehand and subscribed to.

**Review by IljaDaderko — COMMENTED** [review]
Good stuff! I left few comments, I think biggest change that's needed is related to those data flows I mentioned and subscribing to them.

Once that's changed I'll test ui out 👍

File: packages/ui/src/utils/UiHelpers.ts (hunk 7 above)
Comment (crypblizz8) [reply packages/ui/src/utils/UiHelpers.ts]: Removed + rebuilt with one small `tinycolors` library !

File: package.json [hunk 9]
```diff
@@ -25,18 +25,19 @@
...
-    "@typescript-eslint/parser": "5.38.0",
-    "@rollup/plugin-node-resolve": "14.1.0",
     "@rollup/plugin-commonjs": "22.0.2",
     "@rollup/plugin-json": "4.1.0",
+    "@rollup/plugin-node-resolve": "14.1.0",
+    "@typescript-eslint/eslint-plugin": "5.38.0",
+    "@typescript-eslint/parser": "5.38.0",
     "eslint": "8.23.1",
     "eslint-config-prettier": "8.5.0",
+    "lerna": "5.5.2",
```
Comment (crypblizz8) [inline package.json:35]: Think it was just my prettier re-order. Happy to revert

**Comment by crypblizz8** [comment]
### Overview of whole PR
- Create `AccountButton`
- Create `AccountView`
- Create `zorb-ens component`
- Altered data flow to be higher order in the lifecycle
- Added new SVGs
- Added ```tinycolor2``` package

### Recent Changes from last PR check
DataFlow
- This was done by moving up onChange events higher in the order in ```wagmiHelpers``` with ```getBalance``` and ```getENSAvatar``` with a Promise.all
- This will provide initial state data
- In scenario of the change of accounts (I.e. Account 1 --> Account 2 on Metamask), onConnectorChange and sets the relevant state
- New data appears on `AccountButton` and `AccountView` (regardless if model is open)

SVG
- Reworked `BACK_ICON` and old `CHEVRON` into just `CHEVRON`
- Kept `CLIPBOARD_BLUE` as it is a necessary SVG with the blue circle. Can optimize later

Zora
- Installed on library (```npm i tinycolor2```) for Zorb Colors
- Disabled TS-lint on the document
- Used the rest from the Zora `web-component-package`

Learnings
- Make this PR smaller next time.
- Consider the higher order data flow from beginning.

File: packages/ui/src/utils/Zorb.ts [hunk 10]
```diff
@@ -0,0 +1,209 @@
+/* eslint-disable @typescript-eslint/restrict-template-expressions */
```
Comment (IljaDaderko) [inline packages/ui/src/utils/Zorb.ts:1]: Really don't like how much bloat this adds to modal for what it acheives. Let's drop Zorb and all of it's dependencies and instead simplify this in the following way:

There are plenty of ideas of how to generate random color from a given string outlined here:
https://stackoverflow.com/questions/3426404/create-a-hexadecimal-colour-based-on-a-string-with-javascript

Let's grab user's address and split it in 2 parts. Use those parts to create 2 random colors and just style div with them so it looks similar to a zorb. i.e. highlight gradient, body gradient and one of the colors used as border.

File: chains/ethereum/src/utils/wagmiHelpers.ts [hunk 11]
```diff
@@ -18,8 +18,49 @@ export function getClient() {
   return client
 }
 
-function onConnectorChange(event: ConnectorData) {
-  if (event.account) AccountCtrl.setAddress(event.account)
+async function getBalance(account: string) {
+  const opts: GetBalanceOpts = {
```
Comment (IljaDaderko) [inline chains/ethereum/src/utils/wagmiHelpers.ts:22]:
```suggestion
  const opts = {
```

Let typescript do the work

File: chains/ethereum/src/utils/wagmiHelpers.ts [hunk 12]
```diff
@@ -18,8 +18,49 @@ export function getClient() {
...
+  // eslint-disable-next-line line-comment-position, no-inline-comments, no-warning-comments
+  // Todo: Check / Might be setting state twice..?
+  AccountCtrl.setBalance(balance)
+
+  return balance
+}
+
+async function getENSAvatar(account: string) {
+  try {
+    const opts: FetchEnsAvatarOpts = {
```
Comment (IljaDaderko) [inline chains/ethereum/src/utils/wagmiHelpers.ts:38]:
```suggestion
    const opts = {
```

File: chains/ethereum/src/utils/wagmiHelpers.ts [hunk 13]
```diff
@@ -18,8 +18,49 @@ export function getClient() {
...
+async function getBalance(account: string) {
+  const opts: GetBalanceOpts = {
+    addressOrName: account,
+    chainId: `${NAMESPACE}:${AccountCtrl.state.chainId}`,
+    formatUnits: 'ether'
+  }
+
+  const balance = await ClientCtrl.ethereum().fetchBalance(opts)
+  // eslint-disable-next-line line-comment-position, no-inline-comments, no-warning-comments
+  // Todo: Check / Might be setting state twice..?
```
Comment (IljaDaderko) [inline chains/ethereum/src/utils/wagmiHelpers.ts:30]: This needs to be figured out before merging and comments removed

File: chains/ethereum/src/utils/wagmiHelpers.ts [hunk 14]
```diff
@@ -18,8 +18,49 @@ export function getClient() {
...
+    const ensAvatar = await ClientCtrl.ethereum().fetchEnsAvatar(opts)
+    if (ensAvatar !== undefined) {
+      AccountCtrl.setEnsAvatar(ensAvatar)
+
+      return ensAvatar
+    }
+
+    return ''
+  } catch (error) {
+    // eslint-disable-next-line no-console
```
Comment (IljaDaderko) [inline chains/ethereum/src/utils/wagmiHelpers.ts:51]: Show error toast instead of log

File: chains/ethereum/src/utils/wagmiHelpers.ts [hunk 15]
```diff
@@ -18,8 +18,49 @@ export function getClient() {
...
+    console.log('getENSAvatar error', error)
+
+    return ''
+  }
+}
+
+async function onConnectorChange(event: ConnectorData) {
+  if (event.account) {
+    AccountCtrl.setAddress(event.account)
+    await getBalance(event.account)
```
Comment (IljaDaderko) [inline chains/ethereum/src/utils/wagmiHelpers.ts:61]: Promise.all these

File: package.json [hunk 16]
```diff
@@ -25,18 +25,19 @@
...
     "prettier": "2.7.1",
-    "typescript": "4.8.3",
     "rollup": "2.79.0",
     "rollup-plugin-esbuild": "4.10.1",
     "rollup-plugin-minify-html-literals": "1.2.6",
-    "lerna": "5.5.2"
+    "typescript": "4.8.3",
+    "tinycolor2": "^1.4.2"
```
Comment (IljaDaderko) [inline package.json:41]: `tinycolor2` needs to be removed once comment about zorb is addressed

File: packages/ui/package.json [hunk 17]
```diff
@@ -24,6 +24,7 @@
   },
   "devDependencies": {
     "@types/qrcode": "1.5.0",
+    "@types/tinycolor2": "^1.4.3",
```
Comment (IljaDaderko) [inline packages/ui/package.json:27]: Needs to be removed once comment about Zorb is addressed

File: packages/ui/src/components/w3m-zorb-ens-image/index.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
+import { AccountCtrl } from '@web3modal/core'
```
Comment (IljaDaderko) [inline packages/ui/src/components/w3m-zorb-ens-image/index.ts:1]: File can be simplified once Zorb comment is addressed

File: packages/ui/src/partials/w3m-account-button/index.ts [hunk 19]
```diff
@@ -1,16 +1,78 @@
...
+    this.balance = AccountCtrl.state.balance
+    this.ensAvatar = AccountCtrl.state.ensAvatar
+
+    this.unsubscribe = AccountCtrl.subscribe(() => {
+      this.address = AccountCtrl.state.address
+      this.ensAvatar = AccountCtrl.state.ensAvatar
+      this.balance = AccountCtrl.state.balance
+    })
+
+    return () => this.unsubscribe
```
Comment (IljaDaderko) [inline packages/ui/src/partials/w3m-account-button/index.ts:43]: Why is this return needed here? You set `this.unsubscribe` correctly, but never unsubscribe actually. need to add

```ts
public disconnectedCallback() {
    this.unsubscribe?.()
  }
```

File: packages/ui/src/partials/w3m-account-button/styles.ts [hunk 20]
```diff
@@ -60,12 +116,32 @@ export function dynamicStyles() {
...
+    //   fill: ${foreground.inverse};
```
Comment (IljaDaderko) [inline packages/ui/src/partials/w3m-account-button/styles.ts:120]: Commented code

File: packages/ui/src/utils/Svgs.ts [hunk 21]
```diff
@@ -452,6 +452,76 @@ export const DEFI_IMG = svg`
...
+    <path d="M17.665 22.988v6.048l7.608-10.414-7.608 4.366Z" fill="#585F5F"/>
+    <path d="M17.665 29.036v-6.048l-7.603-4.366 7.603 10.414Z" fill="#9EA9A9"/>
+    <path d="m17.665 21.589 7.603-4.368-7.603-3.36v7.728Z" fill="#272A2A"/>
+    <path d="m10.062 17.22 7.602 4.369v-7.727l-7.603 3.359Z" fill="#585F5F"/>
+    <path d="M15.488 1.856a4.357 4.357 0 0 1 4.357 0l9.866 5.695a4.357 4.357 0 0 1 2.178 3.774v11.391c0 1.557-.83 2.995-2.178 3.774l-9.866 5.695a4.357 4.357 0 0 1-4.357 0L5.623 26.49a4.357 4.357 0 0 1-2.179-3.774V11.325c0-1.557.83-2.995 2.179-3.774l9.865-5.695Z" stroke="#000" stroke-opacity=".1"/>
+  </g>
+  </svg>
+`
+
+export const CLIPBOARD_BLUE = svg`
```
Comment (IljaDaderko) [inline packages/ui/src/utils/Svgs.ts:493]: All icons should be white. Their color is set in css via path fill.

File: packages/ui/src/views/w3m-account-view/index.ts [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+    this.balance = AccountCtrl.state.balance
+    this.ensAvatar = AccountCtrl.state.ensAvatar
+
+    this.unsubscribe = AccountCtrl.subscribe(() => {
+      this.address = AccountCtrl.state.address
+      this.ensAvatar = AccountCtrl.state.ensAvatar
+      this.balance = AccountCtrl.state.balance
+    })
+
+    return () => this.unsubscribe
```
Comment (IljaDaderko) [inline packages/ui/src/views/w3m-account-view/index.ts:42]: Same as above, this return is not needed and proper unsubscribing needs to be done in appropriate lit lifecycle method.

File: packages/ui/src/views/w3m-account-view/index.ts [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  private async copyClipboard() {
```
Comment (IljaDaderko) [inline packages/ui/src/views/w3m-account-view/index.ts:45]: Show successful copy toast message here

File: packages/ui/src/views/w3m-account-view/index.ts [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+    await navigator.clipboard.writeText(this.address)
+  }
+
+  private onDisconnect() {
+    ConnectModalCtrl.closeModal()
+    ClientCtrl.ethereum().disconnect()
+  }
+
+  // -- render ------------------------------------------------------- //
+  // eslint-disable-next-line no-warning-comments
```
Comment (IljaDaderko) [inline packages/ui/src/views/w3m-account-view/index.ts:55]: Needs to be addressed before merge

File: packages/ui/src/views/w3m-account-view/index.ts [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+            </div>
+            <div class="w3m-connected-container">
+              <div>${CONNECTED_INDICATOR}</div>
+              <w3m-text variant="small-normal" color="secondary">Connected</w3m-text>
+            </div>
+            </div>
+          </div>
+        </div>
+
+        <div class="w3m-account-divider"></div>
```
Comment (IljaDaderko) [inline packages/ui/src/views/w3m-account-view/index.ts:80]: Use css border instead of extra element

File: packages/ui/src/views/w3m-account-view/index.ts [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+
+  // -- render ------------------------------------------------------- //
+  // eslint-disable-next-line no-warning-comments
+  // ToDo: Optimize CLIPBOARD_BLUE and DISCONNECT ICON with blueBackground
+
+  protected render() {
+    return html`
+      ${dynamicStyles()}
+
+      <div>
```
Comment (IljaDaderko) [inline packages/ui/src/views/w3m-account-view/index.ts:62]: Is this outer div actually needed?

**Review by IljaDaderko — CHANGES_REQUESTED** [review]

**Comment by IljaDaderko** [comment]
All items marked red need to become components i.e

- [ ] `w3m-address` needs to to use existing text component, and utilise format address function
- [ ] `w3m-balance` needs to use existing text component and expose prop for token icon
- [ ] `w3m-round-button` needs to allow for icon and onClick action. Also add hover effect to it as with other buttons

Screenshot 2022-09-23 at 09 41 12

File: packages/ui/src/utils/UiHelpers.ts [hunk 27]
```diff
@@ -66,6 +66,14 @@ export async function preloadImage(src: string) {
   })
 }
 
+export function roundBalance(address: string) {
+  return parseFloat(address).toFixed(2)
+}
+
+export function formatAddress(address: string) {
+  return `${address.substring(0, 5)}...${address.slice(-5)}`
+}
```
Comment (devceline) [inline packages/ui/src/utils/UiHelpers.ts:75]: I feel like this should have a flexible range instead of the fixed `5` but that's a nitpick

Comment (tunnckoCore) [reply packages/ui/src/utils/UiHelpers.ts:75]: It should be first 4 w/o the `0x` and the last 4.

Comment (crypblizz8) [reply packages/ui/src/utils/UiHelpers.ts:75]: Great suggestion, will set with first 4 and last 4 as default + add in an optional range. @devceline @tunnckoCore

File: packages/ui/src/utils/Svgs.ts (hunk 6 above)
Comment (crypblizz8) [reply packages/ui/src/utils/Svgs.ts:455]: Fixed.

**Review by crypblizz8 — COMMENTED** [review]
Have addressed the following in the latest changes

Major
- Zorb rewrite. Tried a few methods here. 1) The ENS Version 2) Gradient div one that was suggested. 3) Own adaptation. The ENS one was too large. The Gradient one didn't end up as pretty and using HSL is a much better outcome. Created own adaptation of the Zorbs + ENS + learnings from Gradient Div attempt. 
- Check if double setting state?
- Fix the Unsubscribe

- `w3m-address `component
- `w3m-balance` component
- `w3m-round-button` component

Minor
- Typescript for getOpts 
- Use css border instead of extra element 
- Toast message ERROR 
- Toast message for Clipboard.
- Is this outer div actually needed? 
- Address formatting
- Test Mobile View + Mobile Edge Cases
